README.md
<!-- TITLE/ -->

# get-current-line

<!-- /TITLE -->

<!-- BADGES/ -->

<span class="badge-githubworkflow"><a href="https://github.com/bevry/get-current-line/actions?query=workflow%3Abevry" title="View the status of this project's GitHub Workflow: bevry"><img src="https://github.com/bevry/get-current-line/workflows/bevry/badge.svg" alt="Status of the GitHub Workflow: bevry" /></a></span>
<span class="badge-npmversion"><a href="https://npmjs.org/package/get-current-line" title="View this project on NPM"><img src="https://img.shields.io/npm/v/get-current-line.svg" alt="NPM version" /></a></span>
<span class="badge-npmdownloads"><a href="https://npmjs.org/package/get-current-line" title="View this project on NPM"><img src="https://img.shields.io/npm/dm/get-current-line.svg" alt="NPM downloads" /></a></span>
<br class="badge-separator" />
<span class="badge-githubsponsors"><a href="https://github.com/sponsors/balupton" title="Donate to this project using GitHub Sponsors"><img src="https://img.shields.io/badge/github-donate-yellow.svg" alt="GitHub Sponsors donate button" /></a></span>
<span class="badge-thanksdev"><a href="https://thanks.dev/u/gh/bevry" title="Donate to this project using ThanksDev"><img src="https://img.shields.io/badge/thanksdev-donate-yellow.svg" alt="ThanksDev donate button" /></a></span>
<span class="badge-patreon"><a href="https://patreon.com/bevry" title="Donate to this project using Patreon"><img src="https://img.shields.io/badge/patreon-donate-yellow.svg" alt="Patreon donate button" /></a></span>
<span class="badge-liberapay"><a href="https://liberapay.com/bevry" title="Donate to this project using Liberapay"><img src="https://img.shields.io/badge/liberapay-donate-yellow.svg" alt="Liberapay donate button" /></a></span>
<span class="badge-buymeacoffee"><a href="https://buymeacoffee.com/balupton" title="Donate to this project using Buy Me A Coffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg" alt="Buy Me A Coffee donate button" /></a></span>
<span class="badge-opencollective"><a href="https://opencollective.com/bevry" title="Donate to this project using Open Collective"><img src="https://img.shields.io/badge/open%20collective-donate-yellow.svg" alt="Open Collective donate button" /></a></span>
<span class="badge-crypto"><a href="https://bevry.me/crypto" title="Donate to this project using Cryptocurrency"><img src="https://img.shields.io/badge/crypto-donate-yellow.svg" alt="crypto donate button" /></a></span>
<span class="badge-paypal"><a href="https://bevry.me/paypal" title="Donate to this project using Paypal"><img src="https://img.shields.io/badge/paypal-donate-yellow.svg" alt="PayPal donate button" /></a></span>
<br class="badge-separator" />
<span class="badge-discord"><a href="https://discord.gg/nQuXddV7VP" title="Join this project's community on Discord"><img src="https://img.shields.io/discord/1147436445783560193?logo=discord&amp;label=discord" alt="Discord server badge" /></a></span>
<span class="badge-twitch"><a href="https://www.twitch.tv/balupton" title="Join this project's community on Twitch"><img src="https://img.shields.io/twitch/status/balupton?logo=twitch" alt="Twitch community badge" /></a></span>

<!-- /BADGES -->

<!-- DESCRIPTION/ -->

Get the current line number of the executing file and method

<!-- /DESCRIPTION -->


## Usage

[Complete API Documentation.](http://master.get-current-line.bevry.surge.sh/docs/)

```javascript
import getCurrentLine from 'get-current-line'
console.log(
    getCurrentLine(/* optional offset information in case the caller is wrapped */)
)
```

```json
{
    "line": "3",
    "method": "Object.<anonymous>",
    "file": "location/to/the/calling/file.js"
}
```

If you wish to have something that fetches the current file and directory paths, such as a cross-enviroment compatible `__filename` and `__dirname` you can use the [`filedirname` package.](https://github.com/bevry/filedirname).

<!-- INSTALL/ -->

## Install

### [npm](https://npmjs.com "npm is a package manager for javascript")

-   Install: `npm install --save get-current-line`
-   Import: `import pkg from ('get-current-line')`
-   Require: `const pkg = require('get-current-line').default`

### [Deno](https://deno.land "Deno is a secure runtime for JavaScript and TypeScript, it is an alternative for Node.js")

``` typescript
import pkg from 'https://unpkg.com/get-current-line@^7.4.0/edition-deno/index.ts'
```
### [Skypack](https://www.skypack.dev "Skypack is a JavaScript Delivery Network for modern web apps")

``` html
<script type="module">
    import pkg from '//cdn.skypack.dev/get-current-line@^7.4.0'
</script>
```
### [unpkg](https://unpkg.com "unpkg is a fast, global content delivery network for everything on npm")

``` html
<script type="module">
    import pkg from '//unpkg.com/get-current-line@^7.4.0'
</script>
```
### [jspm](https://jspm.io "Native ES Modules CDN")

``` html
<script type="module">
    import pkg from '//dev.jspm.io/get-current-line@7.4.0'
</script>
```
### [Editions](https://editions.bevry.me "Editions are the best way to produce and consume packages you care about.")

This package is published with the following editions:
-   `get-current-line` aliases `get-current-line/index.cjs` which uses the [Editions Autoloader](https://github.com/bevry/editions "You can use the Editions Autoloader to autoload the appropriate edition for your consumers environment") to automatically select the correct edition for the consumer's environment
-   `get-current-line/source/index.ts` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") source code with [Import](https://babeljs.io/docs/learn-es2015/#modules "ECMAScript Modules") for modules
-   `get-current-line/edition-browsers/index.js` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") compiled against [ES2022](https://en.wikipedia.org/wiki/ES2022 "ECMAScript 2022") for web browsers with [Import](https://babeljs.io/docs/learn-es2015/#modules "ECMAScript Modules") for modules
-   `get-current-line/edition-es2022/index.js` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") compiled against [ES2022](https://en.wikipedia.org/wiki/ES2022 "ECMAScript 2022") for [Node.js](https://nodejs.org "Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine") 6 || 8 || 10 || 12 || 14 || 16 || 18 || 20 || 21 with [Require](https://nodejs.org/dist/latest-v5.x/docs/api/modules.html "Node/CJS Modules") for modules
-   `get-current-line/edition-es5/index.js` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") compiled against ES5 for [Node.js](https://nodejs.org "Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine") 4 || 6 || 8 || 10 || 12 || 14 || 16 || 18 || 20 || 21 with [Require](https://nodejs.org/dist/latest-v5.x/docs/api/modules.html "Node/CJS Modules") for modules
-   `get-current-line/edition-es2022-esm/index.js` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") compiled against [ES2022](https://en.wikipedia.org/wiki/ES2022 "ECMAScript 2022") for [Node.js](https://nodejs.org "Node.js is a JavaScript runtime built on Chrome's V8 JavaScript engine") 12 || 14 || 16 || 18 || 20 || 21 with [Import](https://babeljs.io/docs/learn-es2015/#modules "ECMAScript Modules") for modules
-   `get-current-line/edition-types/index.d.ts` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") compiled Types with [Import](https://babeljs.io/docs/learn-es2015/#modules "ECMAScript Modules") for modules
-   `get-current-line/edition-deno/index.ts` is [TypeScript](https://www.typescriptlang.org/ "TypeScript is a typed superset of JavaScript that compiles to plain JavaScript.") source code made to be compatible with [Deno](https://deno.land "Deno is a secure runtime for JavaScript and TypeScript, it is an alternative to Node.js")

<!-- /INSTALL -->

<!-- HISTORY/ -->

## History

[Discover the release history by heading on over to the `HISTORY.md` file.](https://github.com/bevry/get-current-line/blob/HEAD/HISTORY.md#files)

<!-- /HISTORY -->

<!-- BACKERS/ -->

## Backers

### Code

[Discover how to contribute via the `CONTRIBUTING.md` file.](https://github.com/bevry/get-current-line/blob/HEAD/CONTRIBUTING.md#files)

#### Authors

-   [Benjamin Lupton](https://balupton.com) — Accelerating collaborative wisdom.

#### Maintainers

-   [Benjamin Lupton](https://balupton.com) — Accelerating collaborative wisdom.

#### Contributors

-   [Benjamin Lupton](https://github.com/balupton) — [view contributions](https://github.com/bevry/get-current-line/commits?author=balupton "View the GitHub contributions of Benjamin Lupton on repository bevry/get-current-line")

### Finances

<span class="badge-githubsponsors"><a href="https://github.com/sponsors/balupton" title="Donate to this project using GitHub Sponsors"><img src="https://img.shields.io/badge/github-donate-yellow.svg" alt="GitHub Sponsors donate button" /></a></span>
<span class="badge-thanksdev"><a href="https://thanks.dev/u/gh/bevry" title="Donate to this project using ThanksDev"><img src="https://img.shields.io/badge/thanksdev-donate-yellow.svg" alt="ThanksDev donate button" /></a></span>
<span class="badge-patreon"><a href="https://patreon.com/bevry" title="Donate to this project using Patreon"><img src="https://img.shields.io/badge/patreon-donate-yellow.svg" alt="Patreon donate button" /></a></span>
<span class="badge-liberapay"><a href="https://liberapay.com/bevry" title="Donate to this project using Liberapay"><img src="https://img.shields.io/badge/liberapay-donate-yellow.svg" alt="Liberapay donate button" /></a></span>
<span class="badge-buymeacoffee"><a href="https://buymeacoffee.com/balupton" title="Donate to this project using Buy Me A Coffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-yellow.svg" alt="Buy Me A Coffee donate button" /></a></span>
<span class="badge-opencollective"><a href="https://opencollective.com/bevry" title="Donate to this project using Open Collective"><img src="https://img.shields.io/badge/open%20collective-donate-yellow.svg" alt="Open Collective donate button" /></a></span>
<span class="badge-crypto"><a href="https://bevry.me/crypto" title="Donate to this project using Cryptocurrency"><img src="https://img.shields.io/badge/crypto-donate-yellow.svg" alt="crypto donate button" /></a></span>
<span class="badge-paypal"><a href="https://bevry.me/paypal" title="Donate to this project using Paypal"><img src="https://img.shields.io/badge/paypal-donate-yellow.svg" alt="PayPal donate button" /></a></span>

#### Sponsors

-   [Andrew Nesbitt](https://nesbitt.io) — Software engineer and researcher
-   [Balsa](https://balsa.com) — We're Balsa, and we're building tools for builders.
-   [Codecov](https://codecov.io) — Empower developers with tools to improve code quality and testing.
-   [Poonacha Medappa](https://poonachamedappa.com)
-   [Rob Morris](https://github.com/Rob-Morris)
-   [Sentry](https://sentry.io) — Real-time crash reporting for your web apps, mobile apps, and games.
-   [Syntax](https://syntax.fm) — Syntax Podcast

#### Donors

-   [Andrew Nesbitt](https://nesbitt.io)
-   [Armen Mkrtchian](https://mogoni.dev)
-   [Balsa](https://balsa.com)
-   [Chad](https://opencollective.com/chad8)
-   [Codecov](https://codecov.io)
-   [dr.dimitru](https://veliovgroup.com)
-   [Elliott Ditman](https://elliottditman.com)
-   [entroniq](https://gitlab.com/entroniq)
-   [GitHub](https://github.com/about)
-   [Hunter Beast](https://cryptoquick.com)
-   [Jean-Luc Geering](https://github.com/jlgeering)
-   [Michael Duane Mooring](https://mdm.cc)
-   [Michael Harry Scepaniak](https://michaelscepaniak.com)
-   [Mohammed Shah](https://github.com/smashah)
-   [Mr. Henry](https://mrhenry.be)
-   [Nermal](https://arjunaditya.vercel.app)
-   [Pleo](https://pleo.io)
-   [Poonacha Medappa](https://poonachamedappa.com)
-   [Rob Morris](https://github.com/Rob-Morris)
-   [Robert de Forest](https://github.com/rdeforest)
-   [Sentry](https://sentry.io)
-   [ServieJS](https://github.com/serviejs)
-   [Skunk Team](https://skunk.team)
-   [Syntax](https://syntax.fm)
-   [WriterJohnBuck](https://github.com/WriterJohnBuck)

<!-- /BACKERS -->

<!-- LICENSE/ -->

## License

Unless stated otherwise all works are:

-   Copyright &copy; [Benjamin Lupton](https://balupton.com)

and licensed under:

-   [Artistic License 2.0](http://spdx.org/licenses/Artistic-2.0.html)

<!-- /LICENSE -->


HISTORY.md
# History

## v7.4.0 2023 December 30

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)
-   Thank you to the sponsors: [Andrew Nesbitt](https://nesbitt.io), [Balsa](https://balsa.com), [Codecov](https://codecov.io), [Poonacha Medappa](https://poonachamedappa.com), [Rob Morris](https://github.com/Rob-Morris), [Sentry](https://sentry.io), [Syntax](https://syntax.fm)

## v7.3.0 2023 December 27

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)
-   Thank you to the sponsors: [Andrew Nesbitt](https://nesbitt.io), [Balsa](https://balsa.com), [Codecov](https://codecov.io/), [Poonacha Medappa](https://poonachamedappa.com), [Rob Morris](https://github.com/Rob-Morris), [Sentry](https://sentry.io), [Syntax](https://syntax.fm)

## v7.2.0 2023 December 6

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v7.1.0 2023 November 25

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v7.0.0 2023 November 24

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)
-   Minimum required Node.js version changed from `node: >=10` to `node: >=4` adapting to ecosystem changes

## v6.10.0 2023 November 21

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.9.0 2023 November 15

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.8.0 2023 November 13

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.7.0 2023 November 1

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)
-   Updated license from [`MIT`](http://spdx.org/licenses/MIT.html) to [`Artistic-2.0`](http://spdx.org/licenses/Artistic-2.0.html)

## v6.6.0 2021 July 30

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.5.0 2021 July 29

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.4.0 2021 July 28

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.3.0 2020 October 29

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.2.0 2020 September 4

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.1.0 2020 September 4

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v6.0.0 2020 September 3

-   Now compatible with a wider range of environments, including web browsers
-   Optimised the fetching of location information from frames
-   Introduced `getFileFromError` and `getLocationFromError` for a wider range of compatibility and usability
-   Introduced `immediate` property to the offset interface
-   Possible Breaking Changes:
    -   Introduced `char` property to the location interface
    -   Renamed `LineInfo` to `Location`
    -   Rneamed `LineOffset` to `Offset`
-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.16.0 2020 August 18

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.15.0 2020 August 18

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.14.0 2020 August 4

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.13.0 2020 July 22

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.12.0 2020 July 22

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.11.0 2020 July 22

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.10.0 2020 June 25

-   Updated dependencies, [base files](https://github.com/bevry/base), and [editions](https://editions.bevry.me) using [boundation](https://github.com/bevry/boundation)

## v5.9.0 2020 June 21


source/test.ts
import { equal } from 'assert-helpers'
import kava from 'kava'
import getCurrentLine, { Offset } from './index.js'

function wrapper(opts?: Offset) {
	return getCurrentLine(opts)
}

function wrapMethod() {
	return wrapper({ method: 'wrapMethod' })
}

function wrapLine() {
	return wrapper({ method: /wrapline/i })
}

kava.suite('get-current-line', function (suite, test) {
	test('success', function named() {
		const { line, method, file } = getCurrentLine()
		equal(file.includes('test'), true, `file [${file}] includes test`)
		equal(method.includes('named'), true, `method [${method}] includes name`)
		const secondLine = getCurrentLine().line
		equal(
			line < secondLine,
			true,
			`second line [${secondLine}] was later than the first line [${line}]`
		)
	})
	test('wrapper', function named() {
		const a = wrapper()
		const b = wrapper()
		equal(a.file.includes('test'), true, `file [${a.file}] includes test`)
		equal(
			a.method.includes('wrapper'),
			true,
			`method [${a.method}] includes wrapper`
		)
		equal(
			a.line,
			b.line,
			'both lines were the same due to the lack of offset in the wrapper'
		)
	})
	test('wrapMethod', function named() {
		const { line, method, file } = wrapMethod()
		equal(file.includes('test'), true, `file [${file}] includes test`)
		equal(method.includes('named'), true, `method [${method}] includes name`)
		const secondLine = wrapMethod().line
		equal(
			line < secondLine,
			true,
			`second line [${secondLine}] was later than the first line [${line}]`
		)
	})
	test('wrapLine', function named() {
		const { line, method, file } = wrapLine()
		equal(file.includes('test'), true, `file [${file}] includes test`)
		equal(method.includes('named'), true, `method [${method}] includes name`)
		const secondLine = wrapLine().line
		equal(
			line < secondLine,
			true,
			`second line [${secondLine}] was later than the first line [${line}]`
		)
	})
})


source/index.ts
/** The combination of location information about the line that was executing at the time */
export interface Location {
	/** the location of the line that was executing at the time */
	line: number
	/** the location of the character that was executing at the time */
	char: number
	/** the method name that was executing at the time */
	method: string
	/** the file path that was executing at the time */
	file: string
}

/**
 * If provided, continue skipping until:
 *
 * 1. The file or method is found
 * 2. Once found, will continue until neither the file nor method are found anymore
 * 3. Once exited, the frame offset will then apply
 *
 * If you wish to capture the found method or the file, combine them with `frames: -1` or `immediate: true`.
 *
 * If you wish for more customisation than this, create an issue requesting passing a custom skip handler function, as more variance to this interface is too much customisation complexity.
 */
export interface Offset {
	/**
	 * if provided, continue until a method containing or matching this string is exited
	 * if provided alongside a file, will continue until neither the file nor method are found
	 * this allows file and method to act as fallbacks for each other, such that if one is not found, it doesn't skip everything
	 */
	method?: RegExp | string | null
	/**
	 * if provided, continue until a file containing or matching this string is exited
	 * if provided alongside a method, will continue until neither the file nor method are found
	 * this allows file and method to act as fallbacks for each other, such that if one is not found, it doesn't skip everything
	 */
	file?: RegExp | string | null
	/**
	 * once we have satisfied the found condition, if any, then apply this index offset to the frames
	 * e.g. 1 would mean next frame, and -1 would mean the previous frame
	 * Use -1 to go back to the found method or file
	 */
	frames?: number
	/**
	 * once we have satisfied the found condition, should we apply the frame offset immediately, or wait until the found condition has exited
	 */
	immediate?: boolean
}

/**
 * For an error instance, return its stack frames as an array.
 */
export function getFramesFromError(error: Error): Array<string> {
	// Create an error
	let stack: Error['stack'] | null, frames: any[]

	// And attempt to retrieve it's stack
	// https://github.com/winstonjs/winston/issues/401#issuecomment-61913086
	try {
		stack = error.stack
	} catch (error1) {
		try {
			// @ts-ignore
			const previous = err.__previous__ || err.__previous
			stack = previous && previous.stack
		} catch (error2) {
			stack = null
		}
	}

	// Handle different stack formats
	if (stack) {
		if (Array.isArray(stack)) {
			frames = Array(stack)
		} else {
			frames = stack.toString().split('\n')
		}
	} else {
		frames = []
	}

	// Parse our frames
	return frames
}

// Compatibility with Node.js versions <10
let frameRegexNamedGroups: RegExp, frameRegexNumberedGroups: RegExp
try {
	frameRegexNamedGroups =
		/\s+at\s(?:(?<method>.+?)\s\()?(?<file>.+?):(?<line>\d+):(?<char>\d+)\)?\s*$/
} catch (error) {
	frameRegexNumberedGroups = /\s+at\s(?:(.+?)\s\()?(.+?):(\d+):(\d+)\)?\s*$/
}

/**
 * Get the locations from a list of error stack frames.
 */
export function getLocationsFromFrames(frames: Array<string>): Array<Location> {
	// Prepare
	const locations: Array<Location> = []

	// Cycle through the frames
	for (let frame of frames) {
		// ensure each frame is a string
		frame = (frame || '').toString()

		// skip empty frames
		if (frame.length === 0) continue

		// Error
		// at file:///Users/balupton/Projects/active/get-current-line/asd.js:1:13
		// at ModuleJob.run (internal/modules/esm/module_job.js:140:23)
		// at async Loader.import (internal/modules/esm/loader.js:165:24)
		// at async Object.loadESM (internal/process/esm_loader.js:68:5)
		if (frameRegexNamedGroups) {
			const match = frame.match(frameRegexNamedGroups)
			if (match && match.groups) {
				locations.push({
					method: match.groups.method || '',
					file: match.groups.file || '',
					line: Number(match.groups.line),
					char: Number(match.groups.char),
				})
			}
		} else {
			const [match, method, file, line, char] =
				frame.match(frameRegexNumberedGroups) || []
			if (match) {
				locations.push({
					method: method || '',
					file: file || '',
					line: Number(line),
					char: Number(char),
				})
			}
		}
	}

	return locations
}

/**
 * If a location is not found, this is the result that is used.
 */
const failureLocation: Location = {
	line: -1,
	char: -1,
	method: '',
	file: '',
}

/**
 * From a list of locations, get the location that is determined by the offset.
 * If none are found, return the failure location
 */
export function getLocationWithOffset(
	locations: Array<Location>,
	offset: Offset
): Location {
	// Continue
	let found: boolean = !offset.file && !offset.method

	// use while loop so we can skip ahead
	let i = 0
	while (i < locations.length) {
		const location = locations[i]

		// the current location matches the offset
		if (
			(offset.file &&
				(typeof offset.file === 'string'
					? location.file.includes(offset.file)
					: offset.file.test(location.file))) ||
			(offset.method &&
				(typeof offset.method === 'string'
					? location.method.includes(offset.method)
					: offset.method.test(location.method)))
		) {
			// we are found, and we should exit immediatelyg, so return with the frame offset applied
			if (offset.immediate) {
				// apply frame offset
				i += offset.frames || 0
				// and return the result
				return locations[i]
			}
			// otherwise, continue until the found condition has exited
			else {
				found = true
				++i
				continue
			}
		}
		// has been found, and the found condition has exited, so return with the frame offset applied
		else if (found) {
			// apply frame offset
			i += offset.frames || 0
			// and return the result
			return locations[i]
		}
		// nothing has been found yet, so continue until we find the offset
		else {
			++i
			continue
		}
	}

	// return failure
	return failureLocation
}

/**
 * Get each error stack frame's location information.
 */
function getLocationsFromError(error: Error): Array<Location> {
	const frames = getFramesFromError(error)
	return getLocationsFromFrames(frames)
}

/**
 * Get the file path that appears in the stack of the passed error.
 * If no offset is provided, then the first location that has a file path will be used.
 */
export function getFileFromError(
	error: Error,
	offset: Offset = {
		file: /./,
		immediate: true,
	}
): string {
	const locations = getLocationsFromError(error)
	return getLocationWithOffset(locations, offset).file
}

/**
 * Get first determined location information that appears in the stack of the error.
 * If no offset is provided, then the offset used will determine the first location information.
 */
export function getLocationFromError(
	error: Error,
	offset: Offset = {
		immediate: true,
	}
): Location {
	const locations = getLocationsFromError(error)
	return getLocationWithOffset(locations, offset)
}

/**
 * Get the location information about the line that called this method.
 * If no offset is provided, then continue until the caller of the `getCurrentLine` is found.
 * @example Input
 * ``` javascript
 * console.log(getCurrentLine())
 * ```
 * @example Result
 * ``` json
 * {
 * 	"line": "1",
 * 	"char": "12",
 * 	"method": "Object.<anonymous>",
 * 	"file": "/Users/balupton/some-project/calling-file.js"
 * }
 * ```
 */
export default function getCurrentLine(
	offset: Offset = {
		method: 'getCurrentLine',
		frames: 0,
		immediate: false,
	}
): Location {
	return getLocationFromError(new Error(), offset)
}
